src/charm.py

```diff
@@ -115,16 +124,143 @@ def _on_config_changed(self, event: ops.HookEvent) -> None:
+        # If there is secret, then we update the secret and add the secret id to peer data
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        if (
+            peer_relation
+            and not self._has_secrets()
+            and not peer_relation.data[self.app].get(SECRET_KEY)
+        ):
+            logger.debug("Updating peer relation data")
+            peer_relation.data[self.app].update({SECRET_KEY: admin_user.access_token})
+        elif (
+            peer_relation
+            and self._has_secrets()
+            and not peer_relation.data[self.app].get(SECRET_ID)
+        ):
+            logger.debug("Adding secret")
+            secret = self.app.add_secret({SECRET_KEY: admin_user.access_token})
+            peer_relation.data[self.app].update({SECRET_ID: secret.id})
+
+    def _get_admin_access_token(self) -> str:
+        """Get admin access token.
+
+        Returns:
+            admin access token.
+        """
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        assert peer_relation is not None  # nosec
+        if not self._has_secrets():
+            secret_value = peer_relation.data[self.app].get(SECRET_KEY)
+        else:
+            secret_id = peer_relation.data[self.app].get(SECRET_ID)
+            if secret_id:
+                secret = self.model.get_secret(id=secret_id)
+                secret_value = secret.get_content().get(SECRET_KEY)
+        return secret_value
+
+    def _enable_mjolnir(self, event: ops.HookEvent) -> None:
+        """Enable mjolnir service.
+
+        The required steps to enable Mjolnir are:
+         - Get an admin access token.
+         - Check if the MJOLNIR_MEMBERSHIP_ROOM room is created.
+         -- Only users from there will be allowed to join the management room.
+         - Create Mjolnir user or get its access token if already exists.
+         - Create the management room or get its room id if already exists.
+         -- The management room will allow only members of MJOLNIR_MEMBERSHIP_ROOM room to join it.
+         - Make the Mjolnir user admin of this room.
+         - Create the Mjolnir configuration file.
+         - Override Mjolnir user rate limit.
+         - Finally, add Mjolnir pebble layer.
 
         Args:
-            event: Event triggering after pebble is ready.
+            event: Event triggering after config is changed.
         """
-        self.change_config(event)
+        container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self.unit.status = ops.MaintenanceStatus("Waiting for pebble")
+            return
+        # Not checking if the pebble service is enabled to skip this
+        # in case there is a charm update that changes Mjolnir configuration
+        self.model.unit.status = ops.MaintenanceStatus("Configuring Mjolnir")
+        admin_access_token = self._get_admin_access_token()
+        try:
+            synapse.get_room_id(
```

Comment:
Perhaps this logic would be better placed in the `synapse`package

Reply:
David suggested something similar, it was moved to a Mjolnir module now.